src/DynamoPackages/PackageLoader.cs

```diff
@@ -307,6 +315,15 @@ private void TryLoadPackageIntoLibrary(Package package)
             }
         }
 
+        private void AddToExtensionPackageMap(PackageInfo pkg, IExtension ext)
+        {
+            if (ExtensionPackageMap == null)
+            {
+                this.ExtensionPackageMap = new Dictionary<IExtension, PackageInfo>();
+            }
+            ExtensionPackageMap.TryAdd(ext, pkg);
+        }
+
```

Comment:
instead of all these new data structures, you could just add a new handler to the PythonEngineManager.Instance.AvailableEngines.CollectionChanged event.
When a new engine is added, you could check what already loaded package contains its assembly localtion. THat way you could match the engine to its package.

Reply:
I tried that, but that approach was missing something. When you trigger the CollectionChanged event you get all the engines added but to do what you said you would again have to do Assembly mining using some load context or, file search in package directories.

Reply:
You would only have to iterate through the already loaded packages and match the root directory of each package to the python engine assembly location